src/nvim/eval.c

```diff
@@ -16753,6 +16855,83 @@ static void f_visualmode(typval_T *argvars, typval_T *rettv)
     curbuf->b_visual_mode_eval = NUL;
 }
 
+// watcheradd(dict, key, funcref) function
+static void f_watcheradd(typval_T *argvars, typval_T *rettv)
+{
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  if (argvars[0].v_type != VAR_DICT ||
+      argvars[1].v_type != VAR_STRING ||
+      (argvars[2].v_type != VAR_FUNC && argvars[2].v_type != VAR_STRING)) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
```

Comment:
I find this kind of validation code extremely repetitive and tend to avoid adding too much detail, especially because in a future PR I want to refactor the function declarations to be more like this:

```c
  {"watcheradd",      3, 3, f_watcheradd, false, {VAR_DICT, VAR_STRING, VAR_FUNC | VAR_STRING},
  {"watcherdel",      3, 3, f_watcheradd, false, {VAR_DICT, VAR_STRING, VAR_FUNC | VAR_STRING},
```

Which would allow automatic validation for every new function added. the `false`part is to determine if the function is callable from sandbox, and the VAR_* constant values needed to be changed into powers of 2.